README.md
# cache-manage-plugin

A Webpack plugin used to manage multiply cache Directory.

Manage different code caches based on the hash.


## Usage

1. Install dependency.

`npm install --save-dev @mokahr/cache-manage-plugin`

2. Update webpack config file.

```js
const CacheManage = require('@mokahr/cache-manage-plugin');

const someHash = createHash();
// your webpack config
module.exports = {
  module: {
    rules: [
      {
        loader: 'babel-loader',
        options: {
          cacheDirectory: `node_modules/.cache/babel-loader/${someHash}`,
        },
      },
    ]
  }
  ...
  plugins: [
    ...
    new CacheManage({    // use
      cacheHash: someHash, 
      dependencyCachePaths: [`node_modules/.cache/babel-loader/${someHash}`],
    }), 
  ]
}
```

## Configuration

```js
new CacheManage(options)
```

### [optional] options.cacheRecordPath: string

cache directory, default value is `node_modules/.cache`

### [required] options.cacheHash: string

cacheHash mast  provide;

### [optional] options.maxAge: number

cache max age,  default value is `1 * 24 * 60 * 60 * 1000` one day

### [optional] options.dependencyCachePaths: 

need to manage cache's cacheDirectory list, default value is `[]` 



CHANGELOG.md
Changelog
=========
Notable changes to this project will be documented.

## 1.0.0

Initial commit


CacheManagePlugin.js
"use strict";
var fs = require("fs");
var path = require("path");
var rimraf = require("rimraf");
var CacheManagePlugin = (function () {
    function CacheManagePlugin(options) {
        this.options = options || {};
        this.options.cacheRecordPath = this.options.cacheRecordPath || 'node_modules/.cache';
        this.options.maxAge = this.options.maxAge || (1 * 24 * 60 * 60 * 1000);
        this.options.cacheHash = this.options.cacheHash;
        this.options.dependencyCachePaths = this.options.dependencyCachePaths || [];
        this.recordPath = this.options.cacheRecordPath + 'cacheRecord.json';
        this.apply = this.apply.bind(this);
    }
    CacheManagePlugin.prototype.getCacheRecord = function () {
        var record = {};
        if (fs.existsSync(path.resolve(this.recordPath))) {
            record = JSON.parse(fs.readFileSync(path.resolve(this.recordPath), { encoding: 'utf8' }));
        }
        return record;
    };
    CacheManagePlugin.prototype.removeOutTimeCache = function (record) {
        var _this = this;
        var now = Date.now();
        var needRemoveHashList = [];
        Object.entries(record).forEach(function (_a) {
            var hash = _a[0], time = _a[1];
            if (now - time > _this.options.maxAge) {
                needRemoveHashList.push(hash);
            }
        });
        needRemoveHashList.forEach(function (hash) {
            _this.options.dependencyCachePaths.forEach(function (_path) {
                var fullPath = _path + "/" + hash;
                console.log('remove overdue cache', fullPath);
                delete record[hash];
                if (fs.existsSync(path.resolve(fullPath))) {
                    rimraf.sync(path.resolve(fullPath));
                }
            });
        });
        return record;
    };
    CacheManagePlugin.prototype.updateRecord = function (hash, record) {
        record[hash] = Date.now();
        if (!fs.existsSync(this.options.cacheRecordPath)) {
            fs.mkdirSync(this.options.cacheRecordPath);
        }
        fs.writeFileSync(path.resolve(this.recordPath), JSON.stringify(record), { encoding: 'utf8' });
    };
    CacheManagePlugin.prototype.apply = function (compiler) {
        var _this = this;
        compiler.plugin('entryOption', function () {
            var record = _this.getCacheRecord();
            record = _this.removeOutTimeCache(record);
            _this.updateRecord(_this.options.cacheHash, record);
        });
    };
    return CacheManagePlugin;
}());
module.exports = CacheManagePlugin;


src/CacheManagePlugin.ts
import * as fs from 'fs';
import * as path from 'path';
import * as rimraf from 'rimraf';

interface Options {
  /**
   * Current cache directory
   * @defaultValue ./
   */
  cacheRecordPath?: string;
  /**
   * cache hash
   * @defaultValue node_modules/.cache
   */
  cacheHash?: string;

  /**
   * cache max age 
   * @defaultValue 1 * 24 * 60 * 60 * 1000
   */
  maxAge?: number,

  /**
   * need to manage cache directory
   * @defaultValue []
   */
  dependencyCachePaths?: string[],
}

interface Record {
  [hash: string]: number, 
}

class CacheManagePlugin {
  options: Options;
  recordPath: string;
  constructor(options: Options) {
    this.options = options || {};
    this.options.cacheRecordPath = this.options.cacheRecordPath || 'node_modules/.cache';
    this.options.maxAge = this.options.maxAge || (1 * 24 * 60 * 60 * 1000) 
    this.options.cacheHash = this.options.cacheHash;
    this.options.dependencyCachePaths = this.options.dependencyCachePaths || [];
    this.recordPath = this.options.cacheRecordPath + 'cacheRecord.json';
    this.apply = this.apply.bind(this);
  }

  getCacheRecord() {
    let record: Record = {};
    if (fs.existsSync(path.resolve(this.recordPath))) {
      record = JSON.parse(fs.readFileSync(path.resolve(this.recordPath), { encoding: 'utf8' })) as Record;
    }
    return record; 
  }

  removeOutTimeCache(record: Record) {
    const now = Date.now();
    const needRemoveHashList: string[] = [];
    Object.entries(record).forEach(([hash, time]) => {
      if (now - time > this.options.maxAge) {
        needRemoveHashList.push(hash);
      }
    })
    needRemoveHashList.forEach((hash) => {
      this.options.dependencyCachePaths.forEach((_path) => {
        const fullPath = `${_path}/${hash}`;

        console.log('remove overdue cache', fullPath)
          // 清除无效的record
        delete record[hash]; 
        if (fs.existsSync(path.resolve(fullPath))) {
          rimraf.sync(path.resolve(fullPath));
        }
      })
    })
    return record;
  }

  updateRecord(hash: string, record: Record) {
    record[hash] = Date.now();
    if (!fs.existsSync(this.options.cacheRecordPath)) {
      fs.mkdirSync(this.options.cacheRecordPath);
    }
    fs.writeFileSync(path.resolve(this.recordPath), JSON.stringify(record), { encoding: 'utf8' });
  }

  apply(compiler: any) {
    compiler.plugin('entryOption', () => {
      let record = this.getCacheRecord();
      record = this.removeOutTimeCache(record);
      this.updateRecord(this.options.cacheHash ,record);
    });
  }
}

export = CacheManagePlugin;
